### PR TITLE
Workaround for #93, #100.

### PR DESCRIPTION
**Pull request by barrystaes**
Its ugly and honestly i dont believe this PR is accepted, but i present it nonetheless because it can help save a lot of people (and Contao developers) a lot of frustration when dealing with this PHP bug.

Workaround is for #93 (and dupe #100 by me) and if anyone has better ideas on how to mitigate this, its welcome.

I dont think the actual problem can be mitigated here (unless we dont ever directly call `posix_getpwuid($int)` but that seems harsh), but for sure the error shown by the Check script can be nicer.

Code based on @fritzmg's comments.
Suggestions welcome!

### CODE REVIEW

**Comment by fritzmg** [comment]
It's not really a PHP bug. PHP simply seems to define a disabled function as "non existant", but still declared. Otherwise you could not get a warning when you try to use the function (if the function is "not declared" you'd get an "undeclared function error").

File: check/controller/bootstrap.php [hunk 1]
```diff
@@ -169,8 +169,14 @@ function _($str) {
...
-	function posix_getpwuid($int) {
-		return array('name'=>$int);
+	// Workaround for https://github.com/contao/check/issues/93 (bug in PHP <=5.6.24 see https://bugs.php.net/bug.php?id=69297)
+	if (!in_array('posix_getpwuid', explode(',', ini_get('disable_functions')))) {
+		function posix_getpwuid($int)
+		{
+			return array('name' => $int);
+		}
+	} else {
+		//die("Cant define 'posix_getpwuid' function due to PHP settings in specific PHP versions. <br />Remove it from 'disable_functions' or for details see https://github.com/contao/check/issues/93#issuecomment-246278772");
```
Comment (fritzmg) [inline check/controller/bootstrap.php]: Why did you leave it commented?

Comment (barrystaes) [inline check/controller/bootstrap.php]: Because im not sure whats best here.. its open for discussion.

Comment (fritzmg) [inline check/controller/bootstrap.php]: Ah ok. Btw. it's not specific PHP versions, it's _all_ PHP versions ;).

**Comment by barrystaes** [comment]
Regarding this being a bug or not..

Well if the PHP devs would say "its by design" its hard to argue. But AFAIK they dont, they seem to be finding and fixing inconsistencies in https://bugs.php.net/bug.php?id=69297 so..

**Comment by fritzmg** [comment]
No, that bug is about something else entirely. That bug is about `function_exists` erroneously returning `true` for functions that are in `disable_functions` when using the OPCache.

It has nothing to do with this issue.

**Comment by barrystaes** [comment]
I also did a test where i wrapped all calls in Check script via a dud `posix_getpwuid_dud` so we dont have to try and define it but fail.
https://github.com/barrystaes/check/tree/workaround-issue93-dudtest

This works great, but i'm not sure this would be representative for Contao. I _assume_ this function is called there too..

**Comment by barrystaes** [comment]
Granted, i may not grasp the role OPCache plays here.

I noticed `shell_exec` is in `disable_functions` also, but does not suffer from this problem since the Check script doesnt try to redefine it. It just outputs what function_exists() said.

Then if its not a bug in PHP, what needs fixing? Does the Check script need to use something other than `function_exists`, like `is_callable` ?

**Comment by barrystaes** [comment]
Or do we need wrapper functions like [my dud](https://github.com/barrystaes/check/tree/workaround-issue93-dudtest) above? My choice would be to have PHP allow redefining a disabled function. Seems like thats what they're doing, no?

**Comment by fritzmg** [comment]
No, the problem is, that the Contao Check absolutely needs the `posix_getpwuid` function. It does not matter if you use `function_exists` or `is_callable` or any other function to check whether a function exists. The problem is not about that.

Even if `is_callable` would return __TRUE__ instead of __FALSE__ for `posix_getpwuid` when `posix_getpwuid` is defined in `disable_functions`, you would still end up with _Warnings_, as shown here: https://github.com/contao/check/issues/93#issuecomment-246274439

The Contao Check would need to be changed, so that it actually checks whether `posix_getpwuid` is in `disable_functions` and show that, just as it would show any other problem with the server environment.

**Comment by barrystaes** [comment]
@fritzmg I agree. I was referring to https://bugs.php.net/bug.php?id=69297

**Comment by fritzmg** [comment]
I see. But "they" are not trying to redefine a disabled function there. That bug report is just about an erroneous behaviour with the `function_exists` function when the name of the function is passed directly as a string while having the OPCache enabled and the function in question present in `disable_functions`.

**Comment by barrystaes** [comment]
Interesting.. so `function_exists($var)` works but `function_exists('string')` wont?! Is there a logical reason for that?

Anyway, i decided to see what functions (in my `disable_functions` list) appear in Contao or Check source code. (as downloaded for install)

I am going to allow these functions for now, but also think the Check should test for occurences in `ini_get('disable_functions')` and give a nice red warning.

|	Function	|	Contao 3.5.16	|	Check 10.20	|
|	---	|	---	|	---	|
|	`exec`	|	3	|	2	|
|	`system`	|	0	|	0	|
|	`passthru`	|	1	|	0	|
|	`shell_exec`	|	0	|	C	|
|	`escapeshellarg`	|	2	|	0	|
|	`escapeshellcmd`	|	0	|	0	|
|	`proc_close`	|	1	|	0	|
|	`proc_open`	|	1	|	C	|
|	`dl`	|	0	|	0	|
|	`popen`	|	0	|	0	|
|	`show_source`	|	0	|	0	|
|	`posix_kill`	|	0	|	0	|
|	`posix_mkfifo`	|	0	|	0	|
|	`posix_getpwuid`	|	0	|	1	|
|	`posix_setpgid`	|	0	|	0	|
|	`posix_setsid`	|	0	|	0	|
|	`posix_setuid`	|	0	|	0	|
|	`posix_setgid`	|	0	|	0	|
|	`posix_seteuid`	|	0	|	0	|
|	`posix_setegid`	|	0	|	0	|
|	`posix_uname`	|	0	|	0	|

Legend: 

- C = only a check, not an actual call.
- number = files that call this function

**Comment by fritzmg** [comment]
> Interesting.. so function_exists($var) works but function_exists('string') wont?! Is there a logical reason for that?

There is no reason for that - that's a bug. That's the bug @leofeyer and you referred to. But it doesn't have anything to do with this issue. It's a bug that occurs when you use __OPCache__. In Contao Check's case it would not create a fatal error, it would create a warning.

The fatal error that occurs is the regular occurrence.

**Comment by fritzmg** [comment]
Just to make it more clear - here is everything you need to know why the fatal error occurs in the Contao Check:

1. `function_exists` returns `false` if a function is either _not declared_ at all or _disabled_. It will return `true` otherwise.
2. functions that are _disabled_ are still _declared_.
3. you cannot redeclare functions.

Thus PHP code like the following:
```php
if (!function_exists('posix_getpwuid')) {
	function posix_getpwuid($int) {
		return array('name'=>$int);
	}
}
```
will lead to a fatal error, when the function in question (here `posix_getpwuid`) is disabled.

**Comment by fritzmg** [comment]
> Anyway, i decided to see what functions (in my disable_functions list) appear in Contao or Check source code (as downloaded for install)

Some of these functions are also needed for the Live Update functionality of Contao (e.g. `exec()`, see https://community.contao.org/de/showthread.php?63986-Update-Warnung). Thus the Contao Check should tell you, that you cannot use the Live Update, if these functions (if there are more than just `exec()` - only @leofeyer can know that, or at least I don't) are disabled.

**Comment by leofeyer** [comment]
See #103.

**Pull request closed without merging**